File: internal/store/session_state_events.go
package store

import (
	"time"
)

// SessionStateEvent records a point-in-time snapshot of a Cursor session's
// metadata. A new row is written only when something changes (mode, model,
// context usage, etc.), so the table is a sparse change-log, not a heartbeat.
type SessionStateEvent struct {
	ID                int64
	SessionID         string
	OccurredAt        time.Time
	UnifiedMode       string
	ModelName         string
	ContextTokensUsed int
	ContextTokenLimit int
}

// RecordSessionStateEvent inserts a new session state snapshot.
// Called by the SessionStateWatcher whenever it detects a change.
func RecordSessionStateEvent(e SessionStateEvent) error {
	db := Open()
	_, err := db.Exec(
		`INSERT INTO session_state_events
		    (session_id, occurred_at, unified_mode, model_name, context_tokens_used, context_token_limit)
		 VALUES (?, ?, ?, ?, ?, ?)`,
		e.SessionID,
		e.OccurredAt.UTC().Format(time.RFC3339),
		e.UnifiedMode,
		e.ModelName,
		e.ContextTokensUsed,
		e.ContextTokenLimit,
	)
	return err
}

// GetSessionStateAt returns the most recent session state event for sessionID
// that occurred at or before the given time. This gives the exact mode, model,
// and context state that was active when a specific user prompt was sent.
// Returns nil if no events exist for this session before the given time.
func GetSessionStateAt(sessionID string, at time.Time) (*SessionStateEvent, error) {
	db := Open()
	row := db.QueryRow(`
		SELECT id, session_id, occurred_at, unified_mode, model_name,
		       context_tokens_used, context_token_limit
		FROM session_state_events
		WHERE session_id = ? AND occurred_at <= ?
		ORDER BY occurred_at DESC
		LIMIT 1
	`, sessionID, at.UTC().Format(time.RFC3339))

	var e SessionStateEvent
	var occurredAtStr string
	err := row.Scan(
		&e.ID, &e.SessionID, &occurredAtStr,
		&e.UnifiedMode, &e.ModelName,
		&e.ContextTokensUsed, &e.ContextTokenLimit,
	)
	if err != nil {
		return nil, nil // no state recorded yet, not an error
	}
	e.OccurredAt, _ = time.Parse(time.RFC3339, occurredAtStr)
	return &e, nil
}

File: crates/pcr-core/src/store/session_state_events.rs
//! Session state change log. Mirrors `cli/internal/store/session_state_events.go`.

use anyhow::Result;
use chrono::{DateTime, Utc};
use rusqlite::{params, OptionalExtension};

use crate::store::db::open;

#[derive(Debug, Clone, Default)]
pub struct SessionStateEvent {
    pub id: i64,
    pub session_id: String,
    pub occurred_at: DateTime<Utc>,
    pub unified_mode: String,
    pub model_name: String,
    pub context_tokens_used: i64,
    pub context_token_limit: i64,
}

pub fn record_session_state_event(e: &SessionStateEvent) -> Result<()> {
    let conn = open();
    conn.execute(
        r#"INSERT INTO session_state_events
            (session_id, occurred_at, unified_mode, model_name, context_tokens_used, context_token_limit)
           VALUES (?, ?, ?, ?, ?, ?)"#,
        params![
            e.session_id,
            e.occurred_at.to_rfc3339_opts(chrono::SecondsFormat::Secs, true),
            e.unified_mode,
            e.model_name,
            e.context_tokens_used,
            e.context_token_limit,
        ],
    )?;
    Ok(())
}

pub fn get_session_state_at(
    session_id: &str,
    at: DateTime<Utc>,
) -> Result<Option<SessionStateEvent>> {
    let conn = open();
    let at_s = at.to_rfc3339_opts(chrono::SecondsFormat::Secs, true);
    let row = conn
        .query_row(
            r#"SELECT id, session_id, occurred_at, unified_mode, model_name,
                      context_tokens_used, context_token_limit
               FROM session_state_events
               WHERE session_id = ? AND occurred_at <= ?
               ORDER BY occurred_at DESC
               LIMIT 1"#,
            params![session_id, at_s],
            |r| {
                let occurred_s: String = r.get(2)?;
                let occurred = DateTime::parse_from_rfc3339(&occurred_s)
                    .map(|d| d.with_timezone(&Utc))
                    .unwrap_or_else(|_| Utc::now());
                Ok(SessionStateEvent {
                    id: r.get(0)?,
                    session_id: r.get(1)?,
                    occurred_at: occurred,
                    unified_mode: r.get::<_, Option<String>>(3)?.unwrap_or_default(),
                    model_name: r.get::<_, Option<String>>(4)?.unwrap_or_default(),
                    context_tokens_used: r.get::<_, Option<i64>>(5)?.unwrap_or_default(),
                    context_token_limit: r.get::<_, Option<i64>>(6)?.unwrap_or_default(),
                })
            },
        )
        .optional()?;
    Ok(row)
}
